pyDrop/clustering/PCRData.py

```diff
@@ -0,0 +1,444 @@
+=========================================================================================
+PCRData: class wrapper for most of the plotting and clustering functionality in pyDrop
+"""
+import pandas as pd
+import numpy as np
+import matplotlib.pyplot as plt
+import re
+from sklearn import cluster
+from sklearn.metrics.cluster import rand_score, homogeneity_score, completeness_score, v_measure_score, silhouette_score
+
+import os
+import math
+from scipy import stats
+from scipy.stats import poisson, norm, chi2, cosine, exponnorm, hypsecant, logistic, genlogistic, dweibull, cauchy, exponweib
+
+from pyDrop.exceptions import (PCRDataReadingException,
+                               FeaturesImproperlySpecified,
+                               LabelsImproperlySpecified,
+                               TooManyDimensions,
+                               SupervisedDataRequired,
+                               NumClustersModelValueError,
+                               ImproperClusterModel)
+
+class PCRData:
+    """Loads Data from file or pandas dataframe. Includes options 
+    for column naming which is required if input data does not follow
+    the format ch1, ch2, ... cluster_* (case in-sensitive).
+    Parameters
+    ----------
+    data : pandas.dfame | str
+        pandas dataframe or file path to csv file containing 
+    X_columns : list(str)
+        column names that correspond to the features. Overrides default names
+    y_columns : str
+        column name that corresponds to the labels. Overrides default names
+    Usage
+    -----
+    >>> filepath = "data/3d_assay_0.csv"
+    >>> data = PCRData(filepath, sep=",", header=0)
+    >>> data.plot()
+    PCRData also contains functionality to subsample data if true labels are given for testing purposes:
+    >>> data.subsample_clusters({1: 1.0, 2: 0, 3: 0.5})
+    """
+    def __init__(self, data, X_columns: list=[], y_column: str="", verbose: bool=False, *args, **kwargs):
+        if isinstance(data, pd.DataFrame):
+            pass
+        elif isinstance(data, str):
+            try:
+                data = pd.read_csv(data, *args, **kwargs)
+            except Exception as e:
+                raise PCRDataReadingException(e)
+        else:
+            raise PCRDataReadingException(f"Data must be of the format str or pandas.DataFrame. Type {type(data)} was given.")
+        
+        self.verbose = verbose
+        
+        if X_columns: # explicitly given X columns
+            if any([name not in data.columns for name in X_columns]): # if not all of X_columns appear 
+                ghost_names = [name for name in X_columns if name not in data.columns]
+                raise FeaturesImproperlySpecified(f"The given names do not appear in the data feature columns: {ghost_names}")
+            else:
+                self.X_cols = X_columns
+        else: # if no explicitly given X columns, try to infer based on a standard format ch*
+            regexp = re.compile(r'ch*', re.I)
```

Comment:
```suggestion
            regexp = re.compile(r'^ch*', re.I)
```

Adding the anchor to the start allows you to skip the span check. You could also just use `startswith` in this case I think